File: DBA.java
/*******************************************************************************
 * Copyright (C) 2018 Francois PETITJEAN 
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3 of the License.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/ 

import static java.lang.Math.sqrt;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Random;


/**
 * This toy class show the use of DBA.
 * @author Francois Petitjean
 */
public class DBA{
	static final long serialVersionUID = 1L;

	private final static int NIL = -1;
	private final static int DIAGONAL = 0;
	private final static int LEFT = 1;
	private final static int UP = 2;

	/**
	 * Performs the DBA averaging by first finding the median over a sample, then doing n iterations of the update
	 * @param sequences set of sequences to average
	 */
	public static double[] performDBA(double[][] sequences) {
		
		int maxLength=0;
		for (int i = 0; i < sequences.length; i++) {
			maxLength = Math.max(maxLength,sequences[i].length);
		}
		double[][]costMatrix = new double[maxLength][maxLength];
		int[][]pathMatrix = new int[maxLength][maxLength];
		int medoidIndex = approximateMedoidIndex(sequences,costMatrix);
		double[]center = Arrays.copyOf(sequences[medoidIndex], sequences[medoidIndex].length);
		
		for (int i = 0; i < 15; i++) {
			center = DBAUpdate(center, sequences,costMatrix,pathMatrix);
		}
		return center;
	}
	
	private static int approximateMedoidIndex(double[][] sequences,double[][]mat) {
		/*
		 * we are finding the medoid, as this can take a bit of time, 
		 * if there is more than 50 time series, we sample 50 as possible 
		 * medoid candidates
		 */
		ArrayList<Integer>allIndices = new ArrayList<>();
		for (int i = 0; i < sequences.length; i++) {
			allIndices.add(i);
		}
		Collections.shuffle(allIndices);
		ArrayList<Integer>medianIndices = new ArrayList<>();
		for (int i = 0; i < sequences.length && i<50; i++) {
			medianIndices.add(allIndices.get(i));
		}
		
		int indexMedoid = -1;
		double lowestSoS = Double.MAX_VALUE;
		
		for (int medianCandidateIndex:medianIndices) {
			double[] possibleMedoid = sequences[medianCandidateIndex];
			double tmpSoS = sumOfSquares(possibleMedoid, sequences,mat);
			if (tmpSoS < lowestSoS) {
				indexMedoid = medianCandidateIndex;
				lowestSoS = tmpSoS;
			}
		}
		return indexMedoid;
	}
	
	private static double sumOfSquares(double[]sequence,double[][]sequences,double[][]mat) {
		double sos = 0.0;
		for (int i = 0; i < sequences.length; i++) {
			double dist = DTW(sequence,sequences[i],mat);
			sos += dist*dist;
		}
		return sos;
	}
	
	public static double DTW(double[]S,double []T,double[][]costMatrix) {
		int i, j;
		costMatrix[0][0] = squaredDistance(S[0],T[0]);
		for (i = 1; i < S.length; i++) {
			costMatrix[i][0] = costMatrix[i - 1][0] + squaredDistance(S[i], T[0]);
		}
		for (j = 1; j < T.length; j++) {
			costMatrix[0][j] = costMatrix[0][j - 1]	+ squaredDistance(S[0],T[j]);
		}
		for (i = 1; i < S.length; i++) {
			for (j = 1; j < T.length; j++) {
				costMatrix[i][j] = Min3(costMatrix[i - 1][j - 1],
						costMatrix[i][j - 1], costMatrix[i - 1][j])
						+ squaredDistance(S[i],T[j]);
			}
		}
		
		return sqrt(costMatrix[S.length - 1][T.length - 1]);
	}
	
	private static double[] DBAUpdate(double[] C, double[][] sequences,double[][]costMatrix,int[][]pathMatrix) {
		double[]updatedMean = new double[C.length];
		int[]nElementsForMean= new int[C.length];
		
		int i, j, indiceRes;
		double res = 0.0;
		int centerLength = C.length;
		int seqLength;

		for (double[] T : sequences) {
			seqLength = T.length;

			costMatrix[0][0] = squaredDistance(C[0], T[0]);
			pathMatrix[0][0] = DBA.NIL;

			for (i = 1; i < centerLength; i++) {
				costMatrix[i][0] = costMatrix[i - 1][0] + squaredDistance(C[i], T[0]);
				pathMatrix[i][0] = DBA.UP;
			}
			for (j = 1; j < seqLength; j++) {
				costMatrix[0][j] = costMatrix[0][j - 1] + squaredDistance(T[j], C[0]);
				pathMatrix[0][j] = DBA.LEFT;
			}

			for (i = 1; i < centerLength; i++) {
				for (j = 1; j < seqLength; j++) {
					indiceRes = DBA.ArgMin3(costMatrix[i - 1][j - 1], costMatrix[i][j - 1], costMatrix[i - 1][j]);
					pathMatrix[i][j] = indiceRes;
					switch (indiceRes) {
						case DIAGONAL:
							res = costMatrix[i - 1][j - 1];
							break;
						case LEFT:
							res = costMatrix[i][j - 1];
							break;
						case UP:
							res = costMatrix[i - 1][j];
							break;
					}
					costMatrix[i][j] = res + squaredDistance(C[i], T[j]);
				}
			}

			i = centerLength - 1;
			j = seqLength - 1;

			while(pathMatrix[i][j]!=DBA.NIL) {
				updatedMean[i]+=T[j];
				nElementsForMean[i]++;
				switch (pathMatrix[i][j]) {
					case DIAGONAL:
						i = i - 1;
						j = j - 1;
						break;
					case LEFT:
						j = j - 1;
						break;
					case UP:
						i = i - 1;
						break;
				}
			}
			assert(i!=0 || j!=0);
			updatedMean[i]+=T[j];
			nElementsForMean[i]++;
		}

		for (int t = 0; t < centerLength; t++) {
			updatedMean[t] /= nElementsForMean[t];
		}
		
		return updatedMean;

	}



	private static double Min3(final double a, final double b, final double c) {
		if (a < b) {
			if (a < c) {
				return a;
			} else {
				return c;
			}
		} else {
			if (b < c) {
				return b;
			} else {
				return c;
			}
		}
	}

	private static int ArgMin3(final double a, final double b, final double c) {
		if (a < b) {
			if (a < c) {
				return 0;
			} else {
				return 2;
			}
		} else {
			if (b < c) {
				return 1;
			} else {
				return 2;
			}
		}
	}

	private static double squaredDistance(double a, double b) {
		double diff = a-b;
		return diff*diff;
	}

	public static void main(String [] args){
		Random r = new Random(3071980);
		double [][]sequences = new double[50][];
		for(int i=0;i<sequences.length;i++){
			sequences[i] = new double[20];
			for(int j=0;j<sequences[i].length;j++){
				sequences[i][j] = Math.cos(r.nextDouble()*j/20.0*Math.PI) ;
			}
		}
		double [] averageSequence = performDBA(sequences);
		
		System.out.print("[");
		for(int j=0;j<averageSequence.length;j++){
			System.out.print(averageSequence[j]+" ");
		}
		System.out.println("]");
	}
}

File: DBA.m
%*******************************************************************************
 % Copyright (C) 2013 Francois PETITJEAN, Ioannis PAPARRIZOS
 % This program is free software: you can redistribute it and/or modify
 % it under the terms of the GNU General Public License as published by
 % the Free Software Foundation, version 3 of the License.
 % 
 % This program is distributed in the hope that it will be useful,
 % but WITHOUT ANY WARRANTY; without even the implied warranty of
 % MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 % GNU General Public License for more details.well
 % 
 % You should have received a copy of the GNU General Public License
 % along with this program.  If not, see <http://www.gnu.org/licenses/>.
 %*****************************************************************************/ 

% function average = DBA(sequences)
% 	index=randi(length(sequences),1);
% 	average=repmat(sequences{index},1);
% 	for i=1:15
% 		average=DBA_one_iteration(average,sequences);
% 	end
% end

function average = DBA(sequences)
    average = repmat(sequences{medoidIndex(sequences)},1);
	for i=1:15
		average=DBA_one_iteration(average,sequences);
	end
end

function sos = sumOfSquares(s,sequences)
    sos = 0.0;
    for i=1:length(sequences)
        dist = dtw(s,sequences{i});
        sos = sos + dist * dist;
    end
end

function score = dtw(S,T)
    costM = zeros(length(S),length(T));
    
    costM(1,1) = (S(1)-T(1))^2;
    for i=2:length(S)
        costM(i,1)= costM(i-1,1)+ (S(i)-T(1))^2;
    end
    for i=2:length(T)
        costM(1,i)= costM(1,i-1)+ (S(1)-T(i))^2;
    end
    for i=2:length(S)
        for j=2:length(T)
            costM(i,j)=min(min(costM(i-1,j-1),costM(i,j-1)),costM(i-1,j))+(S(i)-T(j))^2;
        end
    end
    score = sqrt(costM(length(S),length(T)));
end

function index = medoidIndex(sequences) 
	index = -1;
    lowestInertia = Inf;
	for i=1:length(sequences)
        tmpInertia = sumOfSquares(sequences{i},sequences);
        if (tmpInertia < lowestInertia)
            index = i;
            lowestInertia = tmpInertia;
        end
    end
end


function average = DBA_one_iteration(averageS,sequences)

	tupleAssociation = cell (1, size(averageS,2));
	for t=1:size(averageS,2)
		tupleAssociation{t}=[];
	end

	costMatrix = zeros(1000,1000);
	pathMatrix = zeros(1000,1000);

	for k=1:length(sequences)
	    sequence = sequences{k};
	    costMatrix(1,1) = distanceTo(averageS(1),sequence(1));
	    pathMatrix(1,1) = -1;
	    for i=2:size(averageS,2)
		costMatrix(i,1) = costMatrix(i-1,1) + distanceTo(averageS(i),sequence(1));
		pathMatrix(i,1) = 2;
	    end
	    
	    for j=2:size(sequence,2)
		costMatrix(1,j) = costMatrix(1,j-1) + distanceTo(sequence(j),averageS(1));
		pathMatrix(1,j) = 1;
	    end
	    
	    for i=2:size(averageS,2)
		for j=2:size(sequence,2)
		    indiceRes = ArgMin3(costMatrix(i-1,j-1),costMatrix(i,j-1),costMatrix(i-1,j));
		    pathMatrix(i,j)=indiceRes;
		    
		    if indiceRes==0
			res = costMatrix(i-1,j-1);
		    elseif indiceRes==1
			res = costMatrix(i,j-1);
		    elseif indiceRes==2
			res = costMatrix(i-1,j);
		    end
		    
		    costMatrix(i,j) = res + distanceTo(averageS(i),sequence(j));
		    
		end
	    end
	    
	    i=size(averageS,2);
	    j=size(sequence,2);
	    
	    while(true)
		tupleAssociation{i}(end+1) = sequence(j);
		if pathMatrix(i,j)==0
		    i=i-1;
		    j=j-1;
		elseif pathMatrix(i,j)==1
		    j=j-1;
		elseif pathMatrix(i,j)==2
		    i=i-1;          
		else
		    break;
		end
	    end
	    
	end

	for t=1:size(averageS,2)
	   averageS(t) = mean(tupleAssociation{t});
	end
	   
	average = averageS;

end

function value = ArgMin3(a,b,c)

	if (a<b)
	    if (a<c)
		value=0;
		return;
	    else
		value=2;
		return;
	    end
	else
	    if (b<c)
		value=1;
		return;
	    else
		value=2;
		return;
	    end
	end

end


function dist = distanceTo(a,b)
    dist=(a-b)*(a-b);
end

function ex = test()
    sequences = {};
    sequences{100}=[];
    for i=1:100
        length = randi(100);
        sequences{i}=rand(1,length);
    end
    mean=DBA(sequences);
end
 

	    
